### DISCUSSION TITLE
Small h for task 11

### DISCUSSION THREAD

**Discussion by AprilSweettooth** [Part-II-Computational-Physics/.github#38] [Q&A]
![image](https://user-images.githubusercontent.com/72824407/221928215-572e4eb9-984c-42a7-81d2-365f37de3c61.png)
I am not sure why I get this weird plotting for large N. At small h the spin up and down states should be equal probable so average Mz should be 0 right?

**Comment by JNHallen** [comment]
Yes, the average should be zero, although numerical uncertainty means that my actual result fluctuates around $\pm 10^{-10}$. 

For small $h$ you will have two degenerate eigenvalues. You have to handle this degeneracy to measure any operator in the ground state. I suspect that you may just be using one of the two degenerate eigenvectors.

 The first thing I would check is that you are finding the eigenstates corresponding to the two smallest eigenvalues (and that these are equal). The optional command "which='SA'" can be used in scipy.sparse.linalg.eigsh to ensure that you get the eigenvalues with the smallest amplitude. You then have to combine these in an appropriate way to construct the true ground state.

I would also confirm that the eigenvalues you find for $h=0$ match your expectation ($E=-N$).

**Reply by AprilSweettooth** [reply]
Thanks! I've tried to test the ground state as following:
`
def ExpValueMz(N, h):
    # Your code goes here.
    expVal = []
    for i in range(len(h)):
      psi = spsl.eigsh(LinearHamiltonian(N, h=h[i]), k=1, which='SA')[1]
      print(spsl.eigsh(LinearHamiltonian(N, h=h[i]), k=3, which='SA'))
      e = psi.transpose() @ ActWithmZ(psi.transpose()[0], N)
      expVal.append(e[0])
    return expVal

print('expectation value:', ExpValueMz(3, [0]))[0]`

But then the eigsh gives me same eigenvalue for he first two state

`
(array([-3., -3.,  1.]), array([[ 7.76571479e-01,  6.30029156e-01, -1.03728969e-18],
       [ 0.00000000e+00,  1.38777878e-17,  7.01686286e-01],
       [ 5.55111512e-17, -2.77555756e-17,  3.79524773e-01],
       [ 5.55111512e-17,  0.00000000e+00,  3.39272934e-01],
       [-1.66533454e-16,  2.77555756e-17,  3.99091045e-01],
       [-5.55111512e-17,  0.00000000e+00,  1.84996530e-01],
       [-5.55111512e-17,  5.55111512e-17, -2.34507571e-01],
       [ 6.30029156e-01, -7.76571479e-01, -6.01495692e-18]]))

expectation value: -0.9807360179483583`

I am not sure which one to choose. 
Also, I think the final expectation varies a lot, sometimes -0.9-ish, sometime around 0.6, positive.
Is it because of the fluctuations in built-in functions?

**Reply by JNHallen** [reply]
The problem here is that you are not accounting for the degeneracy properly. The Hamiltonian for $h \ll 1$ has two ground states in terms of our basis states $|n\rangle$. These are 
$$|1\rangle = |\uparrow \uparrow \uparrow ... \uparrow \uparrow \rangle $$
and 
$$|2^N\rangle = |\downarrow \downarrow \downarrow ... \downarrow \downarrow \rangle $$
They have energy $E_0=-N$. 

In terms of our $2^N$ state vector, these two states correspond to 
$$|\phi_1\rangle = (1, 0, 0, ..., 0, 0)^T$$
and 
$$|\phi_2\rangle = (0, 0, 0, ..., 0, 1)^T$$
However, any appropriately normalised linear combination of these two will also be a valid groundstate of the system. Eigsh will find two orthonormal linear combinations of $|\phi_1 \rangle$ and $|\phi_2\rangle$, in the general form 
$$|\psi_1\rangle = \sqrt{a} |\phi_1\rangle + \sqrt{1-a} |\phi_2\rangle$$
and
$$|\psi_2\rangle = \sqrt{1-a} |\phi_1\rangle - \sqrt{a} |\phi_2\rangle$$
for $0\leq a \leq 1$. Annoyingly, it does not find "the same $a$" for every value of $N$ you give it, and this is why you find that the expectation value varies a lot **when you only use one of the two degenerate eigenvectors**. If you look at the two lowest-energy states you are finding for small $h$, you will see that they have this form though. Effectively, the groundstate forms a plane, and cannot be represented by a single vector. To measure the expectation value you should instead take 
$$\langle m^z \rangle = \frac{1}{2} \langle \phi_1| m^z |\phi_1 \rangle + \frac{1}{2} \langle \phi_2| m^z |\phi_2 \rangle$$. 


Finally, keep in mind that you only have two degenerate ground states when $h$ is small.

**Reply by AprilSweettooth** [reply]
Thanks for the explanation. So, if I understand it correctly, I need
`
psi = spsl.eigsh(LinearHamiltonian(N, h=h), k=2, which='SA')[1]
  
hybrid_psi = (1/np.sqrt(2)) * (psi[:,0].transpose() + psi[:,1].transpose())
`
to get $| \phi_1 \rangle$ ? Or alternatively psi[:,0].transpose() - psi[:,1].transpose() to get $| \phi_2 \rangle$.

But I then got this 
![image](https://user-images.githubusercontent.com/72824407/222210757-9913aa9f-972c-45e8-b5d5-4a18995cfcbd.png)

**Reply by JNHallen** [reply]
Sorry what I wrote above is not true. I will edit that comment and try to explain it better.

It looks like you are accounting for a degeneracy also at large $h$, where there is no degeneracy. Make sure to check that you actually have degenerate ground states when you determine your ground state.

**Reply by JNHallen** [reply]
This is perhaps a poorly phrased task. The states $|\psi_1 \rangle$ and $|\psi_2 \rangle$ form an orthonormal basis set for the ground state manifold of ${\cal H}$ (at small $h$). However, unless $a=0$ so that $|\psi_1\rangle = |\phi_1 \rangle$ and $|\psi_2\rangle = |\phi_2 \rangle$, these are not the basis states of $m^z$. There is no unique ground state, and therefore no unique answer to the expectation value $\langle m_z \rangle$. 

With that said, as we do not have any further information about the state it is reasonable to assume that all ground states are equiprobable. In this case, the expectation value will indeed average to zero. Strictly speaking, one could integrate over the full groundstate manifold to show this, but for the sake of this exercise I suggest you just use 
$$\langle m^z \rangle = \frac{1}{2} \langle \phi_1| m^z | \phi_1 \rangle +  \frac{1}{2} \langle \phi_2| m^z | \phi_2 \rangle$$
I.e. that you assume that the two orthonormal states you have found are equiprobable.

**Reply by AprilSweettooth** [reply]
I got something like this,

![image](https://user-images.githubusercontent.com/72824407/222395306-a0dc3289-5899-414d-88b8-a9163ed73328.png)

Yes, I think my states are degenerate. Now the expectation are around 0 but I am not sure why there is still a peak when N=7 and h=1...

Just to check that the above argument does not affect task 10 & 12 right, because both the mx direction and perturbtion 'lifts' the degeneracy.

**Reply by JNHallen** [reply]
The "peak" you are seeing here is just numerical noise. Your expectation values are smaller than $\pm 2\times 10^{-15}$, which is essentially zero. 

In task 10 the ground state of the system is again not fully known when $h$ is small, and the strictly correct thing to do would be to assume you have an equiprobable state in the ground state manifold. However, here we are saved by the fact that $\langle m^x \rangle = 0$ for all possible ground states (which you can quickly check by measuring in the two degenerate orthonormal states you get from eigsh).

In task 12 the whole point is that we lift the degeneracy and get a single ground state, so there you do not have to worry about any of this.
